keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,201 @@
+---
+title: Union types
+authors:
+  - "@apelisse"
+owning-sig: sig-api-machinery
+participating-sigs:
+reviewers:
+approvers:
+- @lavalamp
+editor: TBD
+creation-date: 2019-03-25
+last-updated: 2019-03-25
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+replaces:
+superseded-by:
+---
+
+# Union types
+
+## Summary
+
+Modern data model definitions like OpenAPI v3 and protobuf (versions 2 and 3)
+have a keyword to implement “oneof” or "union". They allow APIs to have a better
+semantic, typically as a way to say “only one of the given fields can be
+set”. We currently have multiple occurrences of this semantic in kubernetes core
+types, at least:
+- DeploymentStrategy is a structure that has a discrminator
+  "DeploymentStrategyType" which decides if "RollingUpate" should be set
+- VolumeSource is a structure that holds the definition of all the possible
+  volume types, only one of them must be set
+
+The problem with the lack of solution is that:
+- The API is implicit, and people don't know how to use it
+- Clients can't know how to deal with that, especially if they can't parse the
+  OpenAPI
+- Server can't understand the user intent and normalize the object properly
+
+## Motivation
+
+Currently, changing a value in an one-of type is difficult because the semantic
+is implicit, which means that nothing can be built to automatically fix unions,
+leading to many bugs and issues:
+- https://github.com/kubernetes/kubernetes/issues/35345
+- https://github.com/kubernetes/kubernetes/issues/24238
+- https://github.com/kubernetes/kubernetes/issues/34292
+- https://github.com/kubernetes/kubernetes/issues/6979
+- https://github.com/kubernetes/kubernetes/issues/33766
+- https://github.com/kubernetes/kubernetes/issues/24198
+- https://github.com/kubernetes/kubernetes/issues/60340
+
+And then, for other people:
+- https://github.com/rancher/rancher/issues/13584
+- https://github.com/helm/charts/pull/12319
+- https://github.com/EnMasseProject/enmasse/pull/1974
+- https://github.com/helm/charts/pull/11546
+- https://github.com/kubernetes/kubernetes/pull/35343
```

Comment:
The lists are great. To this section I'd add that the serverside apply work is what enables us to finally solve this problem.